### PR TITLE
[CIT-471] Add mapper integration tests

### PR DESCRIPTION
**Pull request by rina23q**
Add SM C8Y Mapper integration tests.

Scenarios:
- Mapper publishes SmartREST 114 (supported operation) and 500 (get pending operation) at startup
- Mapper publishes SoftwareList request at startup
- Mapper coverts ThinEdgeJSON SoftwareList response to SmartREST SoftwareList (116) and publishes it.
- Mapper converts SmartREST SoftwareUpdate operation (528) to  ThinEdgeJSON SoftwareUpdate request and publishes it.
- Mapper converts ThinEdgeJSON SoftwareUpdate response to SmartREST SoftwareList (116) and StatusUpdate (501,502,503), and publishes them.

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by abelikt** [comment]
Hi, I cannot really contribute on commenting the code.
A thing that I would have loved to see was a summary in the code that describes what is going to be tested.

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    use std::process::{Child, Stdio};
+    use std::time::Duration;
+    use tedge_config::{ConfigRepository, TEdgeConfig, TEdgeConfigLocation};
+
+    const MQTT_TEST_PORT: u16 = 56666;
+    static SERVER: OnceCell<Child> = OnceCell::new();
+
+    #[tokio::test]
+    #[serial]
+    async fn mapper_publishes_a_software_list_request() {
```
Comment (makr11st) [inline mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:20]: We need to add cfg attribute for `mosquitto-available = []` to run only with the feature enabled for all those tests:
```suggestion
    #[tokio::test]
    #[serial]
    #[cfg_attr(not(feature = "requires-mosquitto"), ignore)]
    
    async fn mapper_publishes_a_software_list_request() {
```

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        // Make sure the server is running.
+        let _start_mosquitto = SERVER.get_or_init(spawn_mqtt_server);
```
Comment (makr11st) [inline mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:22]: For the time being till we have teh solution for spawning the server automatically this should be removed.

**Review by makr11st — COMMENTED** [review]
> Hi, I cannot really contribute on commenting the code.
> A thing that I would have loved to see was a summary in the code that describes what is going to be tested.

Great point, I'll update the PR.

**Review by didier-wenzek — APPROVED** [review]
Approved.

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs (hunk 2 above)
Comment (makr11st) [reply mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:22]: True, thanks for catching it.

**Review by albinsuresh — APPROVED** [review]

File: mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        // Start Mapper
+        let config = create_tedge_config();
+        let _mapper_task = tokio::spawn(async {
+            crate::sm_c8y_mapper::mapper::CumulocitySoftwareManagementMapper::new()
+                .start(config)
+                .await
+        });
+
+        // The first message that arrives on tedge/commands/req/software/list should be software list request.
+        match tokio::time::timeout(Duration::from_millis(1000), received.next()).await {
```
[the remark this answers is not in the document]
Comment (makr11st) [reply mapper/tedge_mapper/src/sm_c8y_mapper/tests.rs:45]: That's a good point, all of that communication is local and theoretically even on a slow machine one second is plenty of time to send and receive data, but I think we may increase it, what timeout do you propose?

**Review by PradeepKiruvale — COMMENTED** [review]
Dint get how the tests run if there is no mqtt broker.

**Comment by makr11st** [comment]
> Dint get how the tests run if there is no mqtt broker.

The tests run only when you provide attribute `mosquitto-available` so eg:
`cargo test -p tedge_mapper --features mosquitto-available`

**Comment by makr11st** [comment]
> Hi, I cannot really contribute on commenting the code.
> A thing that I would have loved to see was a summary in the code that describes what is going to be tested.

I added some more descriptive comments to the tests, let me know if they are good now.

**Review by abelikt — APPROVED** [review]

**Pull request merged**